kb/development/mbedtls-coding-standards.md

```diff
@@ -116,11 +116,17 @@ All public names (functions, variables, types, `enum` constants, macros) must st
     mbedtls_aes_setkey_decrypt()
 ```
 
+Note that all externally linked functions must have a name starting with `mbedtls_` to avoid link-time conflicts, even if they are not declared in a public header. This also applies to global variables (which should be used very sparingly).
+
 Exception: code implementing the PSA crypto API uses the `PSA_` and `psa_` prefixes. This includes official APIs as well as draft APIs that are on the PSA standards track. API extensions which are meant to remain specific to Mbed TLS, and internal functions, should use the `MBEDTLS_/mbedtls_` prefix, however there are many existing cases of using the `PSA_/psa_` prefix.
 
 ### Local names
 
-Static functions and macros that are not in public headers follow the same convention except the initial `MBEDTLS_` or `mbedtls_` prefix: they start directly with the function name.
+Static functions can use the same naming scheme as non-static functions (`mbedtls_MODULE_frobnicate()` or `psa_frobnicate()`). This is convenient if they are made non-static later, for example for testing. They can also have shorter names if it's convenient. Keeping the module name (`MODULE_frobnicate()`) as a prefix makes it makes it slightly easier to follow call stacks across modules, but it is not compulsory.
+
+Macros defined only in the `library` directory should follow the same naming scheme as non-static functions (`MBEDTLS_MODULE_FOO` or `PSA_FOO`).The can use shorter names (omitting `MBEDTLS_`, the module name, or both) if it's convenient. However, avoid names without underscores as some embedded platforms define short macro names in their system headers. Exception: macros defined in headers used by alternative implementations or PSA drivers (including headers that they include) must start with with `MBEDTLS_` or `PSA_`.
+
+All macros must have uppercase names unless they are function-like (expanding to an expression that evaluates each argument exactly once). Using all-uppercase names for preprocessor constants is recommended. Function-like macros, and macros without arguments that expand to function name, usually follow the lowercase naming convention for functions.
```

Comment:
> Function-like macros [...] usually follow the lowercase naming convention for functions

The codebase doesn't seem to agree with this.

It's not exact, but:

```
$ find . -name '*.[ch]' | fgrep -v 3rd | xargs perl -n -e 'print if /define [a-z]\w+\(/' | wc
      10      72     586
$ find . -name '*.[ch]' | fgrep -v 3rd | xargs perl -n -e 'print if /define [A-Z]\w+\(/' | wc
     533    3169   32960
```

Reply:
I've also realised that while the text defines "function-like macros", it defines them to something different to how they are usually understood. e.g. https://www.ibm.com/docs/en/i/7.2?topic=directive-function-like-macros: "Function-like macro definition: An identifier followed by a parameter list in parentheses and the replacement tokens"

Could we have something like "Macros that take parameters must be all-uppercase unless they expand to a direct call to a function, and only evaluate their arguments exactly once"?